ENH: Implement Keyword Aggregation for DataFrame.agg and Series.agg

charlesdong1991 opened the pull request:
- [x] closes #26513, related to #28380 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

This PR basically deals with three issue:
1. implement keyword aggregation for DataFrame.agg
2.  implement keyword aggregation for Series.agg

charlesdong1991 committed:
remove \n from docstring

charlesdong1991 committed:
fix conflicts

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master'

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 committed:
Implement agg for DataFrame

charlesdong1991 committed:
fix conflict

charlesdong1991 committed:
remove unused import

jbrockmendel commented on pandas/core/groupby/generic.py:
```diff
@@ -870,7 +862,7 @@ def aggregate(self, func=None, *args, **kwargs):
 
         func = _maybe_mangle_lambdas(func)
-
+        print("here")
```
pls remove

charlesdong1991 replied:
oops! sorry for this 😅

jbrockmendel commented on pandas/core/groupby/helper.py:
```diff
@@ -0,0 +1,193 @@
+from collections import OrderedDict
```
more informative name than "helper"?

jreback commented on pandas/core/frame.py:
```diff
@@ -6616,16 +6621,53 @@ def _gotitem(
+            func, indexes, order = _normalize_keyword_aggregation(kwargs)
+            reordered_indexes = [
+                pair[0] for pair in sorted(zip(indexes, order), key=lambda t: t[1])
+            ]
```
most of this is already implemented in the SelectionMixin
you should not have to add much code

charlesdong1991 replied:
emm, i think this implementation is sort of post-processing after aggregation from SelectionMixin? no? 
i will try my best to shorten the code anyway! thanks! @jreback

charlesdong1991 committed:
remove print

charlesdong1991 committed:
fix test

charlesdong1991 committed:
fix typo

charlesdong1991 committed:
add keyword agg for series

charlesdong1991 committed:
fix linting

charlesdong1991 committed:
fix PY35 issue

charlesdong1991 committed:
try to fix py35 order issue

charlesdong1991 committed:
test if fixed

charlesdong1991 committed:
test again

charlesdong1991 committed:
simpler code

charlesdong1991 committed:
test py35

charlesdong1991 committed:
fix conflict

charlesdong1991 committed:
test PY35

charlesdong1991 committed:
try to fix py35

charlesdong1991 committed:
find py35 output

charlesdong1991 committed:
test py35

charlesdong1991 committed:
retest py35

charlesdong1991 committed:
retest py35

charlesdong1991 committed:
try to fix py35

charlesdong1991 committed:
try to fix py35

charlesdong1991 committed:
try one more time

charlesdong1991 committed:
fix typo

charlesdong1991 committed:
py35

charlesdong1991 committed:
skip PY35

charlesdong1991 committed:
skip py35

charlesdong1991 committed:
fix typo

charlesdong1991 committed:
skip all py35

charlesdong1991 committed:
skip py35 for series

charlesdong1991 committed:
fix test

charlesdong1991 committed:
skip series py35

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 committed:
merge master and remove helper

charlesdong1991 committed:
remove helper

charlesdong1991 committed:
remove py36

charlesdong1991 committed:
put back imports

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 committed:
avoid circular dependency

charlesdong1991 committed:
fix linting

WillAyd commented:
@charlesdong1991 is this still active? Can you address comment from @jreback ?

charlesdong1991 commented:
thanks for your comment @WillAyd 

i indeed tried to shorten the code based on comment, and this is the revised version, maybe still not simplified enough. (i remembered i tried to remove some codes, but in some cases, especially related to the order of columns, will fail in tests)

How about splitting it into two PRs with one for Series and one for DataFrame? so that this PR can be more quickly reviewed?

jbrockmendel commented on pandas/core/series.py:
```diff
@@ -3800,9 +3800,17 @@ def _gotitem(self, key, ndim, subset=None):
         self._get_axis_number(axis)
+
+        if func is None:
+            # This is due to order issue of dictionary in PY35, e.g. if {"foo"
```
py35 has been dropped; can this be simplified?

charlesdong1991 replied:
yes, `OrderDict` is no longer needed

jbrockmendel commented on pandas/tests/frame/test_apply.py:
```diff
@@ -1367,3 +1367,96 @@ def test_consistency_of_aggregates_of_columns_with_missing_values(self, df, meth
+
+class TestDataFrameNamedAggregate:
+
+    # GH 26513
```
usually these go inside the test func

charlesdong1991 replied:
yeah, sorry, i put this outside was because i wanted it for all functions in the entire class, since they are all coming from the same issue. I changed and put them under each function.

jbrockmendel commented on pandas/tests/frame/test_apply.py:
```diff
@@ -1367,3 +1367,96 @@ def test_consistency_of_aggregates_of_columns_with_missing_values(self, df, meth
+class TestDataFrameNamedAggregate:
+
+    # GH 26513
+    def test_agg_relabel(self):
```
it looks like this is testing several cases of increasing complexity, is that correct?  if so, can you split them into separate tests with informative names

charlesdong1991 replied:
separated!

jbrockmendel commented on pandas/core/frame.py:
```diff
@@ -6588,16 +6588,51 @@ def _gotitem(
+        from pandas.core.groupby.generic import (
+            _is_multi_agg_with_relabel,
+            _maybe_mangle_lambdas,
+            _normalize_keyword_aggregation,
```
importing private functions is sub-optimal.  it looks like all of the stuff that actually _uses_ these functions doesn't rely on `self`.  could a public function be made in groupby.generic (or possibly SelectionMixin per @jreback's comment) that consists of basically 6600-6611?

charlesdong1991 replied:
yes, i made it in the `base` class

and because of this, I think in the follow up PR, I could do a cleaning on the other one in `core/groupby/generic.py`.

jbrockmendel commented:
> so that this PR can be more quickly reviewed?

We appreciate that you've been very patient both here and in other PRs.  You're making some really good contributions, keep it up!

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 commented:
Many thanks for your nice reviews and kind words, I appreciate a lot!! @jbrockmendel 

it's been a while since my last commit and I do need some time to pick up what had done and need to be done following your review. I will try to come up with revised version ASAP, and then pin you! Thanks again!!

charlesdong1991 committed:
code change based on review

charlesdong1991 committed:
remove util

charlesdong1991 committed:
Add docstring

charlesdong1991 committed:
fix circular import

charlesdong1991 commented:
@jreback @jbrockmendel 

I have addressed the both of your comments and update the code as well. Feel free to take a look if you have time! Thanks!

jreback commented on pandas/core/base.py:
```diff
@@ -289,6 +289,42 @@ def _try_aggregate_string_function(self, arg: str, *args, **kwargs):
             f"'{arg}' is not a valid function for '{type(self).__name__}' object"
         )
 
+    def _reconstruct_func(self, func, *args, **kwargs):
```
ok this is adding a lot of code where we already do a lot of this in groupby. I would start by moving 

```
        from pandas.core.groupby.generic import (
            _is_multi_agg_with_relabel,
            _maybe_mangle_lambdas,
            _normalize_keyword_aggregation,
        )
```
all of that to a new module: pandas.core.aggregation (change the current code to use it).

then add reconstruct_func there.

de-privatize all of these (remove the leading underscores).

jreback replied:
this function needs some work, but would like to see a reorg first. (the moving can be a separate pre-cursor PR if its easier)

charlesdong1991 replied:
thanks for your comment! @jreback 

I will do the code change following your suggestions above. and how about doing reorg and making everything work first and collecting reviews from maintainers, and then I will split the reorg into a separate PR?

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 committed:
reorg and deduplicate

charlesdong1991 committed:
remove used imports

charlesdong1991 committed:
fix linting

charlesdong1991 committed:
fix wrong import

charlesdong1991 committed:
fix conflict

charlesdong1991 committed:
isort

charlesdong1991 committed:
fix mypy

charlesdong1991 commented:
Hi, @jreback 

I updated this PR, although it looks quite big, but actually not, it basically includes these steps:
1. Move helper functions to `pandas.core.aggregation.py`
2. Clean the imports
3. Deduplicate the code (the reason to have this `reconstruct_func` method was I found out for frames, the first part of preprocessing is identical, so could use a generic function for them), therefore
4. add the other part of postprocessing/logic for frame and implement logic for series (for frame, the processing seems a bit too much, i recall this was due to some column order issue, espeically for very complex cases(e.g. with lambda, etc), some processing is needed, but this might because back then py35 was there and dict does not preserve order in py35

i would be very looking forward to hearing your feedbacks and hope this could be included in 1.0, thanks!

jreback commented on pandas/core/series.py:
```diff
@@ -3626,9 +3626,14 @@ def _gotitem(self, key, ndim, subset=None):
         self._get_axis_number(axis)
+
+        if func is None:
+            func = dict(kwargs.items())
```
why are you doing this?

charlesdong1991 replied:
this is needed, for cases like this `df.B.agg(foo=sum, bar="min", cat="min")`, this has to be converted so that `func` can be understood.

i removed the line below

jreback replied:
can you add a comment here giving this explanation

jreback replied:
can you update this

charlesdong1991 replied:
yeah, indeed, it's not ideal, it was inspired by https://github.com/pandas-dev/pandas/pull/26399 , and set `func=None` by default in order to use named aggregation where `agg(foo=('a', min), bar=('b', max))` can be parsed later.

since we haven't seen many bug reports about this renaming enhancement in `df.groupby.agg`, so maybe we could let this roll out for `df.agg` and see how it goes?

jreback commented on pandas/tests/frame/test_apply.py:
```diff
@@ -1392,3 +1392,101 @@ def test_consistency_of_aggregates_of_columns_with_missing_values(self, df, meth
         )
+
+
+class TestDataFrameNamedAggregate:
```
can you move these to a new file test_aggregate.py

jreback replied:
and any other agg tests in test_apply as well

jreback replied:
you can move existing tests in a followup (or pre-cursor) PR if easier / more clear

charlesdong1991 replied:
yeah, I would prefer to have a followup PR for moving around, is it okay? the main reason is I added a helper function in `aggregate.py` in this PR, and in case if any changes needed, i would prefer to change it in this PR since it is closely related to `series.agg` and `frame.agg`

jreback commented on pandas/tests/series/test_apply.py:
```diff
@@ -780,3 +780,34 @@ def test_apply_scaler_on_date_time_index_aware_series(self):
         tm.assert_series_equal(result, pd.Series(np.ones(30), dtype="int64"))
+
+
+class TestNamedAggregation:
```
same can you move to a new file

charlesdong1991 replied:
will do in followup PR

jreback replied:
can you do here

jreback commented on pandas/core/frame.py:
```diff
@@ -6618,16 +6619,39 @@ def _gotitem(
 
+        relabeling, func, columns, order = reconstruct_func(func, *args, **kwargs)
+        if relabeling:
+            kwargs = {}
```
why do you need to do this?

charlesdong1991 replied:
right, removed this, not needed

jreback commented on pandas/core/frame.py:
```diff
@@ -6618,16 +6619,39 @@ def _gotitem(
+            # columns, there might be NaN values, so separate these two cases
+            reordered_result = DataFrame(index=columns)
+            idx = 0
+            for col, funcs in func.items():
```
this is extremely inefficient.

instead try appending to a list and concat at the end.

why do you need .values? that is not very idiomatic here, nor is the dropna

jreback replied:
you can also just reorder the indicies and do an indexing selection would be way better

charlesdong1991 replied:
yeah, I tried to reduce half of the code in this loop, i am pretty sure the previous version was due to some order issue.
However, this seems still a bit complex, but this is to keep two things unchanged: 1. the order to first occurrence of columns to be aggregated, 2. the order of the new named column occurrence.

for instance in the example below, i think a correct behaviour should be:
```python
df = pd.DataFrame({"A": [1, 2, 1, 2], "B": [1, 2, 3, 4], "C": [3,4,5,6]})
df.agg(ab=("C", np.max), foo=("A", max), bar=("B", "mean"), cat=("A", "min"), 
            dat=("B", "max"), f=("A", "max"), g=("C", "min"))
```
the column of new aggregated df is : `['C', 'A', 'B']` other than `['A', 'B', 'C']` since it is the order of the first occurrence of them. And the index should be `['ab', 'foo', 'bar', 'cat', 'dat', 'f', 'g']` since it is the order in the agg function.

Any advice to simply the code would be much appreciated!

jreback reviewed (changes requested):
i think its worthile to do a pre-curor where you just move things (code and tests) to core/aggregation.py w/o any other code changes.

charlesdong1991 committed:
Code change based on review

charlesdong1991 committed:
dropna

charlesdong1991 committed:
fix logic

charlesdong1991 committed:
fix logic

charlesdong1991 committed:
remove unused

charlesdong1991 committed:
fix linting

charlesdong1991 committed:
simpler python

charlesdong1991 committed:
fix conflicts

jreback commented on pandas/core/aggregation.py:
```diff
@@ -0,0 +1,225 @@
+
+    Examples
+    --------
+    >>> _normalize_keyword_aggregation({'output': ('input', 'sum')})
```
needs updating here

charlesdong1991 replied:
changed this kind of mistakes for all in docstring #30856

jreback reviewed (changes requested):
can you do a PR w/o any bug fixes, just the code move.

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 committed:
fix conflicts

charlesdong1991 committed:
fix merge error

charlesdong1991 committed:
fixup

charlesdong1991 committed:
fix annotation

charlesdong1991 committed:
fix annotation

charlesdong1991 commented:
I think the PR is small enough this time for a detailed review ^^ @jreback  thanks!

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 commented:
also just in case if you'd have time for a review ^^ @WillAyd  thanks!

WillAyd commented on pandas/core/aggregation.py:
```diff
@@ -5,14 +5,46 @@
 from pandas.core.indexes.api import Index
 
 
+def reconstruct_func(func, *args, **kwargs):
```
Can you annotate this signature? The return is pretty complicated so would be better to explicitly state what that is

charlesdong1991 replied:
ok, I will try to make it optional!

WillAyd commented on pandas/core/frame.py:
```diff
@@ -6741,16 +6742,38 @@ def _gotitem(
     )
     @Appender(_shared_docs["aggregate"])
-    def aggregate(self, func, axis=0, *args, **kwargs):
+    def aggregate(self, func=None, axis=0, *args, **kwargs):
```
Why did this need to change?

charlesdong1991 replied:
this is needed to make it optional, so that e.g. `df.agg(bar=('a', max))` could be handled correctly

jreback commented on pandas/core/aggregation.py:
```diff
@@ -5,14 +5,46 @@
 
+def reconstruct_func(func, *args, **kwargs):
+    """
+    This is the internal function to reconstruct func given if there is relabeling
```
yeah pls format the doc-string, giving examples if possible.

charlesdong1991 replied:
added docstring and examples

jreback commented on pandas/core/aggregation.py:
```diff
@@ -5,14 +5,46 @@
+    if relabeling:
+        func, columns, order = normalize_keyword_aggregation(kwargs)
+
+    elif isinstance(func, list) and len(func) > len(set(func)):
```
can you do these conditions at the top? it makes the logic easier to grok

charlesdong1991 replied:
yeah, moved it up

jreback commented on pandas/core/frame.py:
```diff
@@ -6741,16 +6742,38 @@ def _gotitem(
         if result is None:
             return self.apply(func, axis=axis, args=args, **kwargs)
+
+        if relabeling:
```
can you make another function to do this, e.g. relable results.

an alternative to having 2 functions, is actually to have `reconstruct_func` return a class which keeps the state

e.g. 

```
reconstructor = reconstruct_func(.....)

# the _aggregate / apply logic

return reconstructor.get_result(result)

charlesdong1991 replied:
nice idea!

I have been thinking about it, and for now, i will go for the first option. Since in generic and frame, there is slightly different manipulation after `_aggregate / apply logic`, and not sure if `resample/rollin` will also be slightly different, so I will keep this in mind and once all nested renaming is done, I will try to come up with a clean PR to optimise it!

jreback commented on pandas/core/frame.py:
```diff
@@ -6741,16 +6742,38 @@ def _gotitem(
+    def aggregate(self, func=None, axis=0, *args, **kwargs):
         axis = self._get_axis_number(axis)
 
+        relabeling, func, columns, order = reconstruct_func(func, *args, **kwargs)
```
is this function also used in groupby? did you already make the appropriate changes there?

charlesdong1991 replied:
yeah, this is exactly the same as part of agg in `groupby/generic.py`, and later should be used for `resample/rolling` agg. Shall I replace the corresponding part in `groupby/generic.py` with this function in this PR? i was supposed to do that in a follow-up to make this PR smaller, but maybe it's better to do it here, will do then!

charlesdong1991 committed:
move code

charlesdong1991 committed:
move it back

charlesdong1991 committed:
fixup

charlesdong1991 committed:
add docstring

charlesdong1991 committed:
add func

charlesdong1991 committed:
isort

charlesdong1991 committed:
fix linting

charlesdong1991 committed:
fix linting

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 commented:
this is not urgent PR, some followup reviews would be appreciated @jreback ^^

jreback commented on pandas/core/frame.py:
```diff
@@ -6767,18 +6768,48 @@ def _gotitem(
+
         return result
 
+    @staticmethod
```
don't put this here, move to pandas/core/aggregation.py we really want to limit the methods in frame/series, esp private implementations (also we barely use staticmethods)

charlesdong1991 replied:
makes more sense! moved!

jreback commented on pandas/tests/frame/test_apply.py:
```diff
@@ -1404,3 +1404,101 @@ def test_consistency_of_aggregates_of_columns_with_missing_values(self, df, meth
             none_in_first_column_result, none_in_second_column_result
         )
+
+
```
can you create a new test file for these, maybe test_apply_relabling.py (alt we can make a pandas/tests/frame/apply/..... subdir)

jreback replied:
ok with *moving* existing functions later, but pls create a new file / subdir now.

charlesdong1991 committed:
code change on JR reviews

charlesdong1991 committed:
move

charlesdong1991 committed:
linting

charlesdong1991 commented:
@jreback since simply creating a file/subdir does not make too much sense, and moving the tests around is not a big mess to this PR in terms of the code change, so let's do it all in one go!

charlesdong1991 committed:
isort

WillAyd commented on pandas/core/aggregation.py:
```diff
@@ -5,14 +5,72 @@
 
 
+def reconstruct_func(
+    func: Optional[Union[List[Any], dict]], *args, **kwargs: dict
```
```suggestion
    func: Optional[Union[List, Dict]], *args, **kwargs
```

charlesdong1991 replied:
changed!

WillAyd commented on pandas/core/aggregation.py:
```diff
@@ -5,14 +5,72 @@
 
+def reconstruct_func(
+    func: Optional[Union[List[Any], dict]], *args, **kwargs: dict
+) -> Tuple[bool, Any, Optional[List[str]], Optional[List[int]]]:
```
```suggestion
) -> Tuple[bool, Optional[Callable], Optional[List[Label]], Optional[List[int]]]:
```

charlesdong1991 replied:
thanks!

charlesdong1991 replied:
emm, seem this isn't correct, i slightly changed a bit, advice are wecome

WillAyd commented on pandas/core/aggregation.py:
```diff
@@ -5,14 +5,72 @@
+    Parameters
+    ----------
+    func: aggregated function
+    **kwargs: dict
```
Can you document what these are used for?

charlesdong1991 replied:
added a short sentence

WillAyd commented on pandas/core/aggregation.py:
```diff
@@ -5,14 +5,72 @@
+    False, "min", None, None
+    """
+    relabeling = func is None and is_multi_agg_with_relabel(**kwargs)
+    columns: Optional[List[str]] = None
```
IIUC you can just remove this line - no need for the assignment to None

charlesdong1991 replied:
emm, no, this still needs to be assigned to `None`, because, if not relabeling, we would like to return `None` for `columns` and `order`

WillAyd commented on pandas/core/aggregation.py:
```diff
@@ -5,14 +5,72 @@
+    """
+    relabeling = func is None and is_multi_agg_with_relabel(**kwargs)
+    columns: Optional[List[str]] = None
+    order: Optional[List[int]] = None
```
Same comment

WillAyd commented on pandas/core/aggregation.py:
```diff
@@ -196,3 +254,23 @@ def maybe_mangle_lambdas(agg_spec: Any) -> Any:
+
+    # The reason is self._aggregate outputs different type of result if
+    # any column is only used once in aggregation
+    mask = isinstance(result, ABCDataFrame) and result.isna().any().any()
```
Why do we need multiple `.any()` calls here?

charlesdong1991 replied:
this is because `result.isna()` is still dataframe, and `.any().any()` will return bool, if only one `any`, will be a series and we dont want it for the mask

WillAyd replied:
I don’t totally understand this - can you provide an example?

jreback replied:
you can use np.any

WillAyd commented on pandas/tests/series/test_apply_relabeling.py:
```diff
@@ -0,0 +1,33 @@
+        expected = df["A"].agg({"foo": "sum"})
+        tm.assert_series_equal(result, expected)
+
+        result = df.B.agg(foo="min", bar="max")
```
Can you use label access instead of dot notation? Applicable in a few places

charlesdong1991 replied:
yeah, changed for all in this file

charlesdong1991 committed:
code change

charlesdong1991 committed:
add docstring

charlesdong1991 committed:
add None back

charlesdong1991 committed:
fix annotation

charlesdong1991 committed:
better annotation

charlesdong1991 committed:
fix annotation

charlesdong1991 committed:
fix annotation

charlesdong1991 committed:
fix linting

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 commented:
any followups? @WillAyd @jreback 

thanks a lot!

WillAyd reviewed (changes requested):
@charlesdong1991 generally I find this still very complex and hard to understand. If there are way to clean up arguments (I.e. not use kwargs) and function returns (I.e. not return multiple values sometimes but not others) would go a long way to improving design and easing the review process

jreback commented:
@charlesdong1991 this got a bit lost, can you merge master and will look

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 commented:
thanks for picking this up back @jreback 

Feel free to provide your opinions, and I will make changes accordingly together with @WillAyd previous reviews

jreback commented on pandas/core/aggregation.py:
```diff
@@ -197,3 +256,23 @@ def maybe_mangle_lambdas(agg_spec: Any) -> Any:
+    # any column is only used once in aggregation
+    mask = isinstance(result, ABCDataFrame) and result.isna().any().any()
+
+    for col, fun in func.items():
```
you are mutating the input argument, which IIUC is a ``DataFrame(index=columns)``; why don't you just create it in this function and return it?

This is a very hard to follow api the way you are doing it.

charlesdong1991 replied:
sorry for late reply @jreback 

yeah, agree, this doesn't look optimal, and input is indeed `DataFrame(index=columns)`, the reason not to create it in this function is `DataFrame` is defined in `frame.py`, if we want to use it, then need to import from `frame.py`, and this function is also gonna be used in `frame.py`, which will result in the conflict imports.

Any suggestions on how to work around it?

charlesdong1991 replied:
i remove the input arugment of `reordered_result` and return a dictionary in this method, and form into Dataframe in `frame.py` instead, is it a bit clearer?

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 committed:
simpler python

charlesdong1991 committed:
simpler python

charlesdong1991 committed:
fixup

charlesdong1991 committed:
simplification

charlesdong1991 committed:
better docs

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 committed:
add docs

charlesdong1991 committed:
focs

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 committed:
fix doc

charlesdong1991 committed:
fixup

charlesdong1991 committed:
fix up

charlesdong1991 committed:
fix doctest

charlesdong1991 committed:
doctest

pep8speaks commented:
Hello @charlesdong1991! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/core/aggregation.py`](https://github.com/pandas-dev/pandas/blob/a7439fe83c6d1171f4fe4f273febec093b4884c4/pandas/core/aggregation.py):

> [Line 67:3](https://github.com/pandas-dev/pandas/blob/a7439fe83c6d1171f4fe4f273febec093b4884c4/pandas/core/aggregation.py#L67): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 70:3](https://github.com/pandas-dev/pandas/blob/a7439fe83c6d1171f4fe4f273febec093b4884c4/pandas/core/aggregation.py#L70): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 74:7](https://github.com/pandas-dev/pandas/blob/a7439fe83c6d1171f4fe4f273febec093b4884c4/pandas/core/aggregation.py#L74): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 77:3](https://github.com/pandas-dev/pandas/blob/a7439fe83c6d1171f4fe4f273febec093b4884c4/pandas/core/aggregation.py#L77): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 83:3](https://github.com/pandas-dev/pandas/blob/a7439fe83c6d1171f4fe4f273febec093b4884c4/pandas/core/aggregation.py#L83): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 271:7](https://github.com/pandas-dev/pandas/blob/a7439fe83c6d1171f4fe4f273febec093b4884c4/pandas/core/aggregation.py#L271): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 274:3](https://github.com/pandas-dev/pandas/blob/a7439fe83c6d1171f4fe4f273febec093b4884c4/pandas/core/aggregation.py#L274): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 278:3](https://github.com/pandas-dev/pandas/blob/a7439fe83c6d1171f4fe4f273febec093b4884c4/pandas/core/aggregation.py#L278): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 287:3](https://github.com/pandas-dev/pandas/blob/a7439fe83c6d1171f4fe4f273febec093b4884c4/pandas/core/aggregation.py#L287): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 288:3](https://github.com/pandas-dev/pandas/blob/a7439fe83c6d1171f4fe4f273febec093b4884c4/pandas/core/aggregation.py#L288): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 292:7](https://github.com/pandas-dev/pandas/blob/a7439fe83c6d1171f4fe4f273febec093b4884c4/pandas/core/aggregation.py#L292): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 306:7](https://github.com/pandas-dev/pandas/blob/a7439fe83c6d1171f4fe4f273febec093b4884c4/pandas/core/aggregation.py#L306): [E114](https://duckduckgo.com/?q=pep8%20E114) indentation is not a multiple of four (comment)
> [Line 307:7](https://github.com/pandas-dev/pandas/blob/a7439fe83c6d1171f4fe4f273febec093b4884c4/pandas/core/aggregation.py#L307): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 309:3](https://github.com/pandas-dev/pandas/blob/a7439fe83c6d1171f4fe4f273febec093b4884c4/pandas/core/aggregation.py#L309): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 310:3](https://github.com/pandas-dev/pandas/blob/a7439fe83c6d1171f4fe4f273febec093b4884c4/pandas/core/aggregation.py#L310): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 314:7](https://github.com/pandas-dev/pandas/blob/a7439fe83c6d1171f4fe4f273febec093b4884c4/pandas/core/aggregation.py#L314): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 committed:
rebuild

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 commented:
maybe let's also try to push this PR once again? thanks! @jreback @WillAyd  ^^

let me briefly summarize since it's been here for a while:
1. added a `reconstruct_func` method which just copied pasted the part from `generic.py`, and this func is used to reconstruct funcs if relabelling.
2. added a `relabel_result` method to reorder the aggregated result and put the right label for them as the final aggregated return. The reason to return a dictionary for this function and then construct the DataFrame afterwards in `frame.py` is to avoid circular imports in `aggregation.py`.
3. removed the part from `generic.py` which becomes `reconstruct_func` to reduce code repeats

Your feedbacks to push it again will be very appreciated and apologies for making it not very readable and hard to review. Thus any suggestions would be very welcome!! thanks again!

WillAyd commented on pandas/core/aggregation.py:
```diff
@@ -5,14 +5,84 @@
 
 
+def reconstruct_func(
+    func: Optional[Union[List, Dict]], **kwargs
```
Why is func an optional List / Dict? Should be a callable, no? If so can you add subtypes to clarify?

WillAyd replied:
Can you clarify this? I am pretty confused reading the annotation versus the docstring and implementation as to what this is

charlesdong1991 replied:
Thanks very much for the question, sorry to miss this one!

And indeed, this is very confusing @WillAyd 

The possible scenarios are:
1. normal aggregations, then `func` could be either str, or Callable, or list of them, or the dictionary of both
2. named aggregation, then `func` will be `None`, and all those information are parsed to `kwargs`

I added typings for `func`, and also add more descriptive comments in the docstring to help understand.

Very sorry for this unclear docstring and functionality. And all your further feedbacks or comments are very welcome and appreciated @WillAyd

WillAyd commented on pandas/core/aggregation.py:
```diff
@@ -196,3 +266,49 @@ def maybe_mangle_lambdas(agg_spec: Any) -> Any:
+
+
+def _relabel_result(
+    result: Union[ABCSeries, ABCDataFrame],
```
You can use FrameOrSeriesUnion from pandas._typing, or maybe even FrameOrSeries if applicable here

jreback replied:
can you update for this

charlesdong1991 replied:
done

WillAyd commented on pandas/core/aggregation.py:
```diff
@@ -196,3 +266,49 @@ def maybe_mangle_lambdas(agg_spec: Any) -> Any:
+    func: Dict[str, List[Union[Callable, str]]],
+    columns: Tuple,
+    order: List[int],
+) -> Dict[str, ABCSeries]:
```
Can you replace ABCSeries with Series? The former is created at runtime so doesn't help with static analysis

jreback replied:
is this possible?

charlesdong1991 replied:
changed

WillAyd reviewed (changes requested):
I think moving in the right direction. Sorry for this seeming slow just pretty thorny code paths

jreback commented on pandas/tests/frame/test_apply_relabeling.py:
```diff
@@ -0,0 +1,103 @@
+import numpy as np
```
can you make a sub-dir pandas/tests/frame/apply/test_relabling.py (and move the test_apply.py)

jreback replied:
can you do this

charlesdong1991 replied:
done

jreback commented on pandas/tests/series/test_apply_relabeling.py:
```diff
@@ -0,0 +1,33 @@
+import pandas as pd
```
same as above, make a /apply/ dir

charlesdong1991 replied:
done

jreback commented on pandas/core/aggregation.py:
```diff
@@ -196,3 +266,49 @@ def maybe_mangle_lambdas(agg_spec: Any) -> Any:
+    mask = isinstance(result, ABCDataFrame) and np.any(result.isna())
+
+    for col, fun in func.items():
+        s = result[col][::-1].dropna() if mask else result[col]
```
can you add some comments here on what you are doing

charlesdong1991 replied:
i changed the parsing logic a bit, and put detailed inline comments above, and also slightly made the test more complex for more complex scenario!

jreback commented on pandas/core/aggregation.py:
```diff
@@ -196,3 +266,49 @@ def maybe_mangle_lambdas(agg_spec: Any) -> Any:
+    for col, fun in func.items():
+        s = result[col][::-1].dropna() if mask else result[col]
+        s.index = reordered_indexes[idx : idx + len(fun)]
+        reordered_result_in_dict[col] = s.reindex(columns)
```
add copy=False on the reindex

charlesdong1991 replied:
added!

jreback commented on pandas/core/aggregation.py:
```diff
@@ -196,3 +266,49 @@ def maybe_mangle_lambdas(agg_spec: Any) -> Any:
+    mask = isinstance(result, ABCDataFrame) and np.any(result.isna())
+
+    for col, fun in func.items():
+        s = result[col][::-1].dropna() if mask else result[col]
```
why are you using `result[col][::-1]` ?

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

jreback commented:
@charlesdong1991 this one has slipped the queue, can you merge master and take another look

charlesdong1991 commented:
sure, will try to address all comments soon

jreback commented on pandas/core/aggregation.py:
```diff
@@ -196,3 +266,49 @@ def maybe_mangle_lambdas(agg_spec: Any) -> Any:
+    reordered_indexes = [
+        pair[0] for pair in sorted(zip(columns, order), key=lambda t: t[1])
+    ]
+    reordered_result_in_dict = {}
```
can you type this dict

charlesdong1991 replied:
added the type

jreback reviewed (changes requested):
ok looks pretty good a few comments if you can address, ping on green.

charlesdong1991 committed:
fixup and resolve conflict and merge master

charlesdong1991 committed:
cleaner code

charlesdong1991 committed:
rename

charlesdong1991 committed:
linting

charlesdong1991 committed:
init

charlesdong1991 committed:
better doc

charlesdong1991 commented on pandas/core/aggregation.py:
```diff
@@ -198,6 +268,79 @@ def maybe_mangle_lambdas(agg_spec: Any) -> Any:
+
+    reorder_mask = len(result.columns) > 1
+    for col, fun in func.items():
+        s = result[col].dropna()
+
+        # In the `_aggregate`, the callable names are obtained and used in `result`, and
+        # these names are ordered alphabetically. e.g.
+        #           C2   C1
+        # <lambda>   1  NaN
+        # amax     NaN  4.0
+        # max      NaN  4.0
+        # sum     18.0  6.0
+        # Therefore, the order of functions for each column could be shuffled
+        # accordingly so need to get the callable name if it is not parsed names, and
+        # reorder the aggregated result for each column.
+        # e.g. if df.agg(c1=("C2", sum), c2=("C2", lambda x: min(x))), correct order is
+        # [sum, <lambda>], but in `result`, it will be [<lambda>, sum], and we need to
+        # reorder so that aggregated values map to their functions regarding the order.
+
+        # However there is only one column being used for aggregation, not need to
+        # reorder since the index is not sorted, and keep as is in `funcs`, e.g.
+        #         A
+        # min   1.0
+        # mean  1.5
+        # mean  1.5
+        if reorder_mask:
+            fun = [
+                com.get_callable_name(f) if not isinstance(f, str) else f for f in fun
+            ]
+            col_idx_order = Index(s.index).get_indexer(fun)
+            s = s[col_idx_order]
+
+        # assign the new user-provided "named aggregation" as index names, and reindex
+        # it based on the whole user-provided names.
+        s.index = reordered_indexes[idx : idx + len(fun)]
+        reordered_result_in_dict[col] = s.reindex(columns, copy=False)
+        idx = idx + len(fun)
+    return reordered_result_in_dict
```
Hi, @WillAyd @jreback 

I think this part brought in quite conversations, and i looked back at it and feel there is still something could improve, so i re-designed the implementation a bit, and added quite detailed comment to explain.

The issue of `result` is, after the `self._aggregate` and `mangle_lamdbas`, the index of results will be obtained from callable name (e.g. `sum` -> `'sum'`, or `np.max` -> `'amax'`), or kept as is for string case (e.g. `'sum'` -> `'sum'`), or a parsed lambda func (e.g. `lambda x: min(x)` -> `'<lambda>'`), and index will be sorted.

so, if the call to function is `df.agg(aa=(C2, sum), ab=(C1, max), ac=(C1, np.max), ad=(C2, lambda x: min(x)), ae=(C1, sum))`, the order of functions is `C1: [max, np.max, sum]` and `C2: [sum, lambda]`, but it will look like something below:
```
             C2   C1
<lambda>      1  NaN
amax        NaN  4.0
max         NaN  4.0
sum        18.0  6.0
```
Therefore, need to find the func name in index, and its position, and do a reordering for `s`, which you see in `reorder_mask` clause. And afterwards, we need to replace those names with the names users provided in the call, e.g. `aa, ab, ae` etc

The reason to have the mask is, if the result is a DataFrame and there is only one column being used for aggregation, e.g. `df.agg(aa=(C1, sum), ab=(C1, mean))`, there will be no re-sort in index, and the order is the same as called in function.

I know this might bring in a lot of confusions, but i have tried many scenarios for corner cases, seems this solution works quite good, and overall this is to achieve 3 things: 
1. the index name of aggregated dataframe has the same order of appearance as in the call, 
2. the column names of aggregated dataframes has the same order of appearance of column names as in the call, 
3. the aggregated result mapping is correct. 

Many thanks for your precious time reviewing this PR and would love to hear your feedbacks! thanks in advance! @WillAyd @jreback

charlesdong1991 replied:
emm, i _think_ CI failure is irrelevant

charlesdong1991 committed:
complex case

charlesdong1991 committed:
linting

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 committed:
resolve conflict and merge master

charlesdong1991 commented:
emm, CI failure seems irrelevant.

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 commented:
some further feedbacks are very welcomed! @jreback @WillAyd

charlesdong1991 committed:
add typing

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 committed:
black

charlesdong1991 committed:
remove line

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 commented:
ping @jreback @WillAyd  for further reviews, thanks in advance!

charlesdong1991 committed:
simplify annotation

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 commented on pandas/core/aggregation.py:
```diff
@@ -5,12 +5,107 @@
+        Union[
+            Union[Callable, str],
+            List[Union[Callable, str]],
+            Dict[Label, Union[Union[Callable, str], List[Union[Callable, str]]]],
```
yeah, it's indeed superfluous, i changed to `Dict[Label, Union[Callable, str, List[Union[Callable, str]]]`

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

jreback commented on pandas/core/aggregation.py:
```diff
@@ -5,12 +5,107 @@
+    func: Optional[
+        Union[
+            Union[Callable, str],
+            List[Union[Callable, str]],
```
crazy signatures, can you create an alias to make this easier to read, *maybe* some of this is Label?

jreback commented on pandas/core/aggregation.py:
```diff
@@ -198,6 +293,79 @@ def maybe_mangle_lambdas(agg_spec: Any) -> Any:
     return mangled_aggspec
 
 
+def _relabel_result(
```
can you make this public (no _) as we are calling from a related module (its not exposed to users at all, just easier to grok)

jreback reviewed (changes requested):
lgtm. very nice, just a minor changes. ping on green.

charlesdong1991 committed:
deprivate relabel_result

jreback reviewed (approved):
lgtm @charlesdong1991 typing cleanings can e a followup

charlesdong1991 commented:
oh, sorry, i am working on the annotations, just haven't pushed the commit @jreback 

will ping you for review once i have simplify annotations!

charlesdong1991 committed:
cleaner annotations

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into nested_renaming_agg

charlesdong1991 committed:
fix import sorting

charlesdong1991 committed:
move defined annotation inside aggregation.py

charlesdong1991 commented:
emm, the errors look unrelated to this PR, I will rebase once there is new commits merged to master.

In the meantime, maybe pls take a look at the annotations so that i could make adjustments if needed? @jreback

charlesdong1991 commented:
ehh?? looks good again somehow? 😅   i thought i saw errors back then

sorry for the noise

jreback commented on pandas/core/aggregation.py:
```diff
@@ -5,12 +5,99 @@
 from pandas.core.indexes.api import Index
+from pandas.core.series import FrameOrSeriesUnion, Series
+
+# types of `func` kwarg for DataFrame.aggregate and Series.aggregate
```
great, we *may* want to think about putting these in _typing, but separate issue / PR (you can just try it if you want), but in a follown.

charlesdong1991 replied:
ahh, okay! i deliberately moved it from _typing to here in commit 05921af since i thought this might be only used in this file.

I will do a PR for moving it to _typing.

jreback commented:
thanks @charlesdong1991 very nice!
